<a href="https://colab.research.google.com/github/DButmeh/-Sales-Prediction-Project/blob/main/Prediction_of_Product_Sales_Project_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Project Overview:
- Load and Inspect Data
- Clean Data
- Exploratory Data Analysis
- Feature Inspection

#  **"Loading Data"**

--------
-----
### **The goal** : understand the properties of products and outlets that play crucial roles in increasing sales

In [ ]:
#Mount google drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# import numpy and pandas
import numpy as np
import pandas as pd

In [ ]:
# Read in the data with Pandas
fpath = "/content/drive/MyDrive/CodingDojo/ project sales predictions/sales_predictions_2023.csv"
df=pd.read_csv(fpath)
# Display the first 5 rows
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


####Q: A1) How many rows are there? ---------------- A:  8523
####Q: A2) How many columns are there?--------------- A: 12
####Q: A3) How many numeric columns?--------------------- A: 5
####Q: A4) How many categorical/string columns are there (dtype='object')? -----A: 7

####3) Are there duplicates? If so, drop any duplicates.
  - No duplications

In [ ]:
df.duplicated().sum()

0

####4) Identify missing values.
  - Outlet_Size / dtype object/ 2410 null values
  - Item_Weight /  dtype float64/ 1463 null values

In [ ]:
null_sums=df.isna().sum()
null_sums

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

####5) Address the missing values by using a placeholder value.
  - A: Address "Outlet_Size" null values by "MISSING" object type
  - A: Address "Item_Weight " null values by (-1) num type

In [ ]:
# Address "Outlet_Size" null values by "MISSING" .
df["Outlet_Size"]=df["Outlet_Size"].fillna("MISSING")
df["Outlet_Size"].isna().sum()


0

In [ ]:
df["Outlet_Size"].value_counts()

Outlet_Size
Medium     2793
MISSING    2410
Small      2388
High        932
Name: count, dtype: int64

In [ ]:
mean=df.groupby("Item_Type")["Item_Weight"].mean()
mean=mean.mean()
mean

12.74329824154449

In [ ]:
df["Item_Weight"]=df["Item_Weight"].fillna(mean)
df["Item_Weight"].isna().sum()


0

##### from 1559 Item_ Identifier we only have the weight of 414 item .**dropping** from the **count** the **null value**

In [ ]:
df.nunique()

Item_Identifier              1559
Item_Weight                   416
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     4
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

## fill  the "Item_Weight "null value  with (-1)

In [ ]:
 mean=df.groupby("Item_Type")["Item_Weight"].mean()
 mean


Item_Type
Baking Goods             12.357684
Breads                   11.608406
Breakfast                12.763448
Canned                   12.379873
Dairy                    13.309938
Frozen Foods             12.847109
Fruits and Vegetables    13.141528
Hard Drinks              11.594870
Health and Hygiene       13.073254
Household                13.278300
Meat                     12.802012
Others                   13.643110
Seafood                  12.591529
Snack Foods              12.944670
Soft Drinks              11.990391
Starchy Foods            13.575502
Name: Item_Weight, dtype: float64

In [ ]:
df['Item_Weight'].isna().sort_values()

0       False
5690    False
5689    False
5688    False
5687    False
        ...  
2833    False
2832    False
2831    False
2845    False
8522    False
Name: Item_Weight, Length: 8523, dtype: bool

#6) Confirm that there are no missing values after addressing them.

In [ ]:
df.isna().sum()


Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

####7) Find and fix any inconsistent categories of data (for example: fix cat, Cat, and cats so that they are consistent).

######Check each string column's value counts for inconsistent values

In [ ]:
#select object columns
col_ob=df.select_dtypes("object").columns
col_ob

Index(['Item_Identifier', 'Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

In [ ]:
#for loop to Check each string column
for col in col_ob:
  print(f"value counts for {col}")
  print(df[col].value_counts())
  print("\n")

value counts for Item_Identifier
Item_Identifier
FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: count, Length: 1559, dtype: int64


value counts for Item_Fat_Content
Item_Fat_Content
Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: count, dtype: int64


value counts for Item_Type
Item_Type
Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: count, dtype: int64


value counts for Outlet_Identifier
Outlet_Identifier
OUT027    935
OUT013

In [ ]:
# Standardize the values in "Item_Fat_Content" column
df["Item_Fat_Content"]=df["Item_Fat_Content"].replace({"Low Fat":"LF",
                                                       "Regular":"Reg",
                                                       "low fat":"LF",
                                                       "reg":"Reg"})
df["Item_Fat_Content"].value_counts()

Item_Fat_Content
LF     5517
Reg    3006
Name: count, dtype: int64

####8) For any numerical columns, obtain the summary statistics of each (min, max, mean).

In [ ]:
# Obtain summary statistics for numeric columns
df.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  8523.000000      8523.000000  8523.000000                8523.000000   
mean     12.838017         0.066132   140.992782                1997.831867   
std       4.226344         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       9.310000         0.026989    93.826500                1987.000000   
50%      12.743298         0.053931   143.012800                1999.000000   
75%      16.000000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

In [ ]:
df.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

In [ ]:
#Name the file
fpath_out = "/content/drive/MyDrive/CodingDojo/ project sales predictions/sales_predictions_2023cl.csv"
df.to_csv(fpath_out, index=False)


In [ ]:
# Read in the data with Pandas
load = "/content/drive/MyDrive/CodingDojo/ project sales predictions/sales_predictions_2023cl.csv"
dfc=pd.read_csv(load)
# Display the first 5 rows
dfc.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30               LF         0.016047   
1           DRC01         5.92              Reg         0.019278   
2           FDN15        17.50               LF         0.016760   
3           FDX07        19.20              Reg         0.000000   
4           NCD19         8.93               LF         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998     MISSING               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [ ]:
dfc.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64